src/main/java/net/neoforged/neoforge/event/ValidityEvent.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.event;
+
+import java.util.Optional;
+import net.minecraft.core.BlockPos;
+import net.minecraft.world.entity.Entity;
+import net.minecraft.world.entity.player.Player;
+import net.minecraft.world.entity.vehicle.ContainerEntity;
+import net.minecraft.world.level.Level;
+import net.minecraft.world.level.block.entity.SignBlockEntity;
+import net.neoforged.neoforge.common.NeoForge;
+import net.neoforged.neoforge.event.entity.player.PlayerEvent;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Event to check if a menu can stay open or if the player has moved too far away from it.
+ * <p>
+ * This is intended for mods that allow a player remote access to blocks and their menus,
+ * but it also can be used to forcibly kick them out of menus.
+ * <p>
+ * You can query the result of the original distance check with {@link #getOriginalValue()}
+ * and set your override with {@link #setResult(boolean)}.
+ * <p>
+ * This base class is never posted (but it can be subscribed to, as it is not abstract),
+ * only one of its subclasses:
+ * <ul>
+ * <li>{@link ValidityEvent.Menu} for all kinds of menus,
+ * <li>{@link ValidityEvent.EntityMenu} for menus attached to an entity, or
+ * <li>{@link ValidityEvent.ContainerEntityMenu} for {@link ContainerEntity}-based menus, or
+ * <li>{@link ValidityEvent.Sign} for the checks done by signs which don't use menus but a Screen.
+ * </ul>
+ * This event is fired on the {@link NeoForge#EVENT_BUS}.
+ */
+public class ValidityEvent extends PlayerEvent {
+    private final double distance;
+    private final boolean originalValue;
+    private Optional<Boolean> result = Optional.empty();
```

Comment:
```suggestion
    private final boolean isOriginallyValid;
    private boolean isValid;
```
No need to use `Optional` for field with setter, the field name should be more self-explanatory.

Reply:
I used the Optional for its nice orElse(), which makes the code that produces the final result more expressive. However, looking at the event again, it seems like I forgot separate getters for the original/override. This means a subscriber cannot distinguish between the original value and something another earlier subscriber set. Oops.